torch/autograd/gradcheck.py

```diff
@@ -201,6 +202,40 @@ def get_analytical_jacobian(input, output, nondet_tol=0.0, grad_out=1.0):
 
     return jacobian, reentrant, correct_grad_sizes, correct_grad_types
 
+def get_analytical_jacobian_fw(fn, input, output):
```

Comment:
To test the correctness of forward mode AD computing the whole Jacobian using finite differences and "analytical" is not needed. It's enough to compute a directional derivative for a random direction (Jacobian-vector product for a random tangent vector) 
```python
v = ... # random vector like input
analytical_vjp = pytorch_vjp(func, input, v)
numerical_vjp = finite_differences_vjp(func, input, v)
torch.allclose(analytical_vjp, numerical_vjp)
```
```python
def finite_differences_vjp(func, inp, v):
"""Jacobian of `func` at `inp` multiplied by `v`"""
    # or using any other more accurate finite difference scheme
    return (func(inp + eps * v) - func(inp)) / (eps * norm(v))
```
So approximately computing JVP by finite difference takes only two function calls. Avoiding computing the whole Jacobian would speed up testing a lot.

Reply:
Yes, That is also mostly true as well for backward mode.
The point here is more that having the full Jacobian is much more useful as an error message :) 

A fast version of gradcheck that uses that strategy you're talking about is in the pipe though and would also be applied to forward AD. But the full Jacobian version will still be kept to ensure good error messages in gradcheck.